File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/offline/ChainVkOfflineNoteProofProvider.java
package org.hyperledger.iroha.android.offline;

/** Offline Note proof provider that proves against a chain-supplied VerifyingKeyBox. */
public final class ChainVkOfflineNoteProofProvider implements OfflineNoteProofProvider {
  private final byte[] vkBoxNorito;

  public ChainVkOfflineNoteProofProvider(final byte[] vkBoxNorito) {
    if (vkBoxNorito == null || vkBoxNorito.length == 0) {
      throw new IllegalArgumentException("vkBoxNorito must not be empty");
    }
    this.vkBoxNorito = vkBoxNorito.clone();
  }

  @Override
  public OfflineNote.RecursiveProof proveAudit(final OfflineNote.AuditBundle audit) {
    final byte[] proofNorito =
        NativeOfflineNoteProver.proveAudit(OfflineNote.encodeAudit(audit), vkBoxNorito);
    return OfflineNote.decodeRecursiveProof(proofNorito);
  }

  @Override
  public OfflineNote.RecursiveProof proveRedeem(final OfflineNote.Redeem redemption) {
    final byte[] proofNorito =
        NativeOfflineNoteProver.proveRedeem(OfflineNote.encodeRedeem(redemption), vkBoxNorito);
    return OfflineNote.decodeRecursiveProof(proofNorito);
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/offline/ChainVkOfflineNoteProofProvider.kt
package org.hyperledger.iroha.sdk.offline

/** Offline Note proof provider that proves against a chain-supplied `VerifyingKeyBox`. */
class ChainVkOfflineNoteProofProvider(vkBoxNorito: ByteArray) : OfflineNoteProofProvider {
    private val vkBoxNorito = vkBoxNorito.copyOf()

    init {
        require(this.vkBoxNorito.isNotEmpty()) { "vkBoxNorito must not be empty" }
    }

    override fun proveAudit(audit: OfflineNote.AuditBundle): OfflineNote.RecursiveProof {
        val proofNorito = NativeOfflineNoteProver.proveAudit(OfflineNote.encodeAudit(audit), vkBoxNorito)
        return OfflineNote.decodeRecursiveProof(proofNorito)
    }

    override fun proveRedeem(redemption: OfflineNote.Redeem): OfflineNote.RecursiveProof {
        val proofNorito = NativeOfflineNoteProver.proveRedeem(OfflineNote.encodeRedeem(redemption), vkBoxNorito)
        return OfflineNote.decodeRecursiveProof(proofNorito)
    }
}
